File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cs
public class Solution {
    
    public string[] map = new string[]
    {
        "",
        "",
        "abc",
        "def",
        "ghi",
        "jkl",
        "mno",
        "pqrs",
        "tuv",
        "wxyz",
        
    };
    
    public IList<string> LetterCombinations(string digits) {
        IList<string> res = new List<string>();
        if(string.IsNullOrEmpty(digits))
            return res;
        Backtrack(res, "", 0, digits);
        return res;
    }
    
    public void Backtrack(IList<string> result, string current, int index, string digits)
    {
        if(index == digits.Length)
        {
            result.Add(current);
            return;
        }
        
        int digit = digits[index] - '0';
        string letters = map[digit];
        
        for(int i=0; i<letters.Length; i++)
        {
            Backtrack(result, current + letters[i], index+1, digits);
        }
    }
}

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:

    unordered_map<char, string> m; 
    
        

    vector<string> letterCombinations(string digits) {

        vector<string> res;
        string temp;

        if(digits == "")
            return res;

        m['2']= "abc";
        m['3']= "def"; 
        m['4']= "ghi"; 
        m['5']= "jkl";
        m['6']= "mno";
        m['7']= "pqrs"; 
        m['8']= "tuv";
        m['9']= "wxyz";

        helper(0, temp, digits, res);
        return res;

    }

    void helper(int idx, string& temp, string digits, vector<string>& res) {
        if(idx == digits.size()) 
        {
            res.push_back(temp);
            return;
        }

        string s = m[digits[idx]];
        for(int i=0; i<s.size(); i++) {
            temp.push_back(s[i]);
            helper(idx+1, temp, digits, res);
            temp.pop_back();
        }
    }
};